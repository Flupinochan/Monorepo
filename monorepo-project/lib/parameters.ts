interface Env {
  account: string;
  region: string;
}

interface CodeCommitParameters {
  repositoryName: string;
  description: string;
}

interface S3Parameters {
  s3Name: string;
}

interface LambdaParameters {
  functionName: string;
  roleName: string;
  logGroupName: string;
  S3Sync_dir_list: string;
}

interface CodePipelineParameters {
  pipelineName: string;
  branchName: string;
  artifactS3Name: string;
  roleName: string;
  actionRoleName: string;
}

interface EventBridgeParameters {
  ruleName: string;
}

interface CodeBuildParameters {
  projectName: string;
  roleName: string;
  logGroupName: string;
}

export class Parameters {
  env: Env = {
    account: "99999999999",
    region: "ap-northeast-3",
  };

  codeCommit: CodeCommitParameters = {
    repositoryName: "Monorepo",
    description: "Monorepo",
  };

  deployTargetS3A: S3Parameters = {
    s3Name: "monorepo-target-example-s3-a",
  };

  deployTargetS3B: S3Parameters = {
    s3Name: "monorepo-target-example-s3-b",
  };

  lambda: LambdaParameters = {
    functionName: "triggerMonorepoPipeline-lambda",
    roleName: "lambdaIAMRole-triggerMonorepoPipeline-lambda",
    logGroupName: "triggerMonorepoPipeline-lambda-logs",
    S3Sync_dir_list:
      "monorepo-target-example-s3-a,monorepo-target-example-s3-b",
  };

  codebuildProject: CodeBuildParameters = {
    projectName: "monorepo-codebuild",
    roleName: "codebuildIAMRole-monorepo",
    logGroupName: "monorepo-codebuild-logs",
  };

  codePipeline: CodePipelineParameters = {
    pipelineName: "monorepo-pipeline",
    branchName: "main",
    artifactS3Name: "monorepo-pipeline-artifact",
    roleName: "pipelineIAMRole-monorepo",
    actionRoleName: "pipelineActionIAMRole-monorepo",
  };

  eventBridge: EventBridgeParameters = {
    ruleName: "triggerMonorepoPipeline-lambda-rule",
  };
}
