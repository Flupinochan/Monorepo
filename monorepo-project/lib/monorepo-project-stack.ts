import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

import * as codecommit from "aws-cdk-lib/aws-codecommit";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as targets from "aws-cdk-lib/aws-events-targets";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as codebuild from "aws-cdk-lib/aws-codebuild";
import * as logs from "aws-cdk-lib/aws-logs";
// import * as ec2 from 'aws-cdk-lib/aws-ec2'; // Snapshot Test用
import path = require("path");
import { Duration } from "aws-cdk-lib";

import { Parameters } from "./parameters";

const param = new Parameters();

export class MonorepoProjectStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // CodeCommit
    const repo = new codecommit.Repository(this, "CodeCommit", {
      repositoryName: param.codeCommit.repositoryName,
      description: param.codeCommit.description,
      code: codecommit.Code.fromZipFile(
        "lib/codecommit/monorepo-target-example-s3.zip",
        param.codePipeline.branchName
      ),
    });

    // Deploy Target S3 A
    const deployS3A = new s3.Bucket(this, "S3_DeployTargetA", {
      bucketName: param.deployTargetS3A.s3Name,
    });

    // Deploy Target S3 B
    const deployS3B = new s3.Bucket(this, "S3_DeployTargetB", {
      bucketName: param.deployTargetS3B.s3Name,
    });

    // Lambda
    const lambdaRole = new iam.Role(this, "Lambda_IAMRole", {
      roleName: param.lambda.roleName,
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        codebuildPolicy: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: ["codecommit:*", "codepipeline:*", "logs:*"],
              resources: ["*"],
            }),
          ],
        }),
      },
    });
    const lambdaLogGroup = new logs.LogGroup(this, "Lambda_CloudWatchLogs", {
      logGroupName: param.lambda.logGroupName,
      retention: logs.RetentionDays.ONE_DAY,
      logGroupClass: logs.LogGroupClass.STANDARD,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    const fn = new lambda.Function(this, "Lambda", {
      functionName: param.lambda.functionName,
      runtime: lambda.Runtime.PYTHON_3_12,
      handler: "index.lambda_handler",
      role: lambdaRole,
      code: lambda.Code.fromAsset(path.join(__dirname, "lambda-code")),
      timeout: Duration.minutes(15),
      logGroup: lambdaLogGroup,
      environment: {
        S3_SYNC_DIR_LIST: param.lambda.S3Sync_dir_list,
        CODEPIPELINE: param.codePipeline.pipelineName,
        LOG_LEVEL: "DEBUG",
      },
    });

    // CodeBuild
    const buildRole = new iam.Role(this, "CodeBuild_IAMRole", {
      roleName: param.codebuildProject.roleName,
      assumedBy: new iam.ServicePrincipal("codebuild.amazonaws.com"),
      inlinePolicies: {
        codebuildPolicy: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "codecommit:*",
                "codebuild:*",
                "codepipeline:*",
                "logs:*",
                "s3:*",
                "iam:*",
              ],
              resources: ["*"],
            }),
          ],
        }),
      },
    });
    const codebuildLogGroup = new logs.LogGroup(
      this,
      "CodeBuild_CloudWatchLogs",
      {
        logGroupName: param.codebuildProject.logGroupName,
        retention: logs.RetentionDays.ONE_DAY,
        logGroupClass: logs.LogGroupClass.STANDARD,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      }
    );
    const codebuildProject = new codebuild.PipelineProject(this, `CodeBuild`, {
      projectName: param.codebuildProject.projectName,
      role: buildRole,
      environment: {
        buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
      },
      logging: {
        cloudWatch: {
          logGroup: codebuildLogGroup,
        },
      },
      buildSpec: codebuild.BuildSpec.fromObject({
        version: "0.2",
        phases: {
          build: {
            commands: [
              "echo ${sourceDirectoryName}",
              "echo ${deployS3BucketName}",
              "aws s3 sync ${sourceDirectoryName} s3://${deployS3BucketName} --delete",
            ],
          },
        },
      }),
    });

    // CodePipeline
    const pipelineRole = new iam.Role(this, "CodePipeline_IAMRole", {
      roleName: param.codePipeline.roleName,
      assumedBy: new iam.ServicePrincipal("codepipeline.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "AWSCodePipeline_FullAccess"
        ),
      ],
    });
    const pipelineActionRole = new iam.Role(
      this,
      "CodePipelineAction_IAMRole",
      {
        roleName: param.codePipeline.actionRoleName,
        assumedBy: new iam.AccountPrincipal(param.env.account),
        inlinePolicies: {
          codepipelineActionPolicy: new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  "codecommit:*",
                  "codebuild:*",
                  "codepipeline:*",
                  "logs:*",
                  "s3:*",
                ],
                resources: ["*"],
              }),
            ],
          }),
        },
      }
    );
    const artifactS3 = new s3.Bucket(this, "S3_CodePipelineArtifact", {
      bucketName: param.codePipeline.artifactS3Name,
    });
    const sourceOutput = new codepipeline.Artifact();
    const sourceDirectoryName = new codepipeline.Variable({
      variableName: "sourceDirectoryName",
      description: "Receive from Lambda",
      defaultValue: "default",
    });
    const deployS3BucketName = new codepipeline.Variable({
      variableName: "deployS3BucketName",
      description: "Receive from Lambda",
      defaultValue: "default",
    });
    const pipeline = new codepipeline.Pipeline(this, "CodePipeline", {
      pipelineName: param.codePipeline.pipelineName,
      artifactBucket: artifactS3,
      pipelineType: codepipeline.PipelineType.V2,
      executionMode: codepipeline.ExecutionMode.QUEUED,
      role: pipelineRole,
      // setting default pipeline valiables
      variables: [sourceDirectoryName, deployS3BucketName],
      stages: [
        {
          stageName: "source",
          actions: [
            new codepipeline_actions.CodeCommitSourceAction({
              actionName: "SourceCodeCommit",
              role: pipelineActionRole,
              repository: repo,
              branch: param.codePipeline.branchName,
              output: sourceOutput,
              runOrder: 1,
              customEventRule: {
                ruleName: param.eventBridge.ruleName,
                eventPattern: {
                  detailType: ["CodeCommit Repository State Change"],
                  resources: [repo.repositoryArn],
                  source: ["aws.codecommit"],
                  detail: {
                    referenceType: ["branch"],
                    event: ["referenceCreated", "referenceUpdated"],
                    referenceName: [param.codePipeline.branchName],
                  },
                },
                target: new targets.LambdaFunction(fn),
              },
            }),
          ],
        },
        {
          stageName: "Build",
          actions: [
            new codepipeline_actions.CodeBuildAction({
              actionName: "BuildS3",
              role: pipelineActionRole,
              input: sourceOutput,
              runOrder: 2,
              project: codebuildProject,
              // CodePipeline変数は、CodeBuild側ではなく、CodePipeline内のBuildステージで設定する必要がある
              environmentVariables: {
                sourceDirectoryName: {
                  value: "#{variables.sourceDirectoryName}",
                  type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                },
                deployS3BucketName: {
                  value: "#{variables.deployS3BucketName}",
                  type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
                },
              },
            }),
          ],
        },
      ],
    });

    // Snapshot Test用
    // const vpc = new ec2.Vpc(this, 'VPC');
  }
}
