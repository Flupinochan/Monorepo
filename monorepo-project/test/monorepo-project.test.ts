import * as cdk from "aws-cdk-lib";
import * as logs from "aws-cdk-lib/aws-logs";
import { Capture, Match, Template } from "aws-cdk-lib/assertions";

import { MonorepoProjectStack } from "../lib/monorepo-project-stack";
import { Parameters } from "../lib/parameters";

// Example
const sum = (x: number, y: number): number => {
  return x + y;
};
describe("Example Test", () => {
  test("sum関数のテスト", () => {
    expect(sum(1, 2)).toBe(3);
  });
});

// MonorepoProjectStack Test
describe("MonorepoProjectStack Test", () => {
  // bin/monorepo-project.ts と同じようにコードを記載する
  // 仮でテスト構築し、その仮で作成したリソースに対して検証テストする (mockしたリソースに対するテスト?)
  // ただし、CloudFormation Templateの書式に合わせる
  const param = new Parameters();
  const app = new cdk.App();
  const stack = new MonorepoProjectStack(app, "MonorepoProjectStack", {
    env: {
      account: param.env.account,
      region: param.env.region,
    },
  });
  const template = Template.fromStack(stack);

  test("AWS::CodeCommit::Repository が1つ作成されていることをテスト", () => {
    template.resourceCountIs("AWS::CodeCommit::Repository", 1);
  });

  test("CodeCommitの名前や説明が正しいかテスト", () => {
    template.hasResourceProperties("AWS::CodeCommit::Repository", {
      RepositoryName: param.codeCommit.repositoryName,
      RepositoryDescription: param.codeCommit.description,
    });
  });

  test("Lambda用CloudWatch Logsのプロパティテスト", () => {
    template.hasResourceProperties("AWS::Logs::LogGroup", {
      LogGroupName: param.lambda.logGroupName,
      RetentionInDays: logs.RetentionDays.ONE_DAY,
      LogGroupClass: logs.LogGroupClass.STANDARD,
    });
  });

  // テストする際は、変換されたCFnテンプレートに対してテストされる
  // そのため、CDKで、roleNameと定義していても、CFnでは、RoleName(Rが大文字)のため、
  // テストでは、RoleNameでテストする
  // cdk synthしたCFnをもとにテストすると良い
  // Match.objectLike() は、いずれかに一致すればTrue
  test("Lambda用IAMRoleテスト", () => {
    template.hasResourceProperties(
      "AWS::IAM::Role",
      Match.objectLike({
        RoleName: param.lambda.roleName,
        Policies: [
          {
            PolicyName: "codebuildPolicy",
            PolicyDocument: {
              Statement: [
                {
                  Effect: "Allow",
                  Action: ["codecommit:*", "codepipeline:*", "logs:*"],
                  Resource: "*",
                },
              ],
            },
          },
        ],
      })
    );
  });

  // Captureの使用例 (ただし、以下はあまり意味がないテスト)
  test("作成されたEventBridge Rule名が、paramで定義したEventBridge Rule名と一致するかテスト", () => {
    const eventBridgeRuleNameCapture = new Capture();
    // 作成されたEventBridge Rule名を取得 (以下は、Nameの値がキャプチャされる)
    template.hasResourceProperties("AWS::Events::Rule", {
      Name: eventBridgeRuleNameCapture,
    });
    // paramで定義したEventBridge Rule名と一致するかテスト
    expect(eventBridgeRuleNameCapture.asString()).toEqual(
      param.eventBridge.ruleName
    );
  });

  // test/__snapshots__/monorepo-project.test.ts.snap に、SnapshotとしてCFnテンプレートが生成される
  // 2度目以降のテストでは、上記CFnテンプレートと比較され、異なればテストが失敗する
  // tsc -wしておくこと! jsファイルから作成されたCFnテンプレートに対して、Snapshotが取得されるため
  // cdk diffは、実際にデプロイされているリソースとのテストのため、異なる
  test("Snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
