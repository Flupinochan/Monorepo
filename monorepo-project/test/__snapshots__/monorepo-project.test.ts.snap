// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonorepoProjectStack Test Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CodeBuild2FDE9E35": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "LogsConfig": {
          "CloudWatchLogs": {
            "GroupName": {
              "Ref": "CodeBuildCloudWatchLogs58D96F43",
            },
            "Status": "ENABLED",
          },
        },
        "Name": "monorepo-codebuild",
        "ServiceRole": {
          "Fn::GetAtt": [
            "CodeBuildIAMRole99B4B757",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "build": {
      "commands": [
        "echo \${sourceDirectoryName}",
        "echo \${deployS3BucketName}",
        "aws s3 sync \${sourceDirectoryName} s3://\${deployS3BucketName} --delete"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "CodeBuildCloudWatchLogs58D96F43": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupClass": "STANDARD",
        "LogGroupName": "monorepo-codebuild-logs",
        "RetentionInDays": 1,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CodeBuildIAMRole99B4B757": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "codecommit:*",
                    "codebuild:*",
                    "codepipeline:*",
                    "logs:*",
                    "s3:*",
                    "iam:*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "codebuildPolicy",
          },
        ],
        "RoleName": "codebuildIAMRole-monorepo",
      },
      "Type": "AWS::IAM::Role",
    },
    "CodeBuildIAMRoleDefaultPolicy59F13077": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodeBuildCloudWatchLogs58D96F43",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:ap-northeast-3:247574246160:log-group:/aws/codebuild/",
                      {
                        "Ref": "CodeBuild2FDE9E35",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:ap-northeast-3:247574246160:log-group:/aws/codebuild/",
                      {
                        "Ref": "CodeBuild2FDE9E35",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:ap-northeast-3:247574246160:report-group/",
                    {
                      "Ref": "CodeBuild2FDE9E35",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3CodePipelineArtifact91067FA4",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3CodePipelineArtifact91067FA4",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodeBuildIAMRoleDefaultPolicy59F13077",
        "Roles": [
          {
            "Ref": "CodeBuildIAMRole99B4B757",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CodeCommit506A881C": {
      "Properties": {
        "Code": {
          "BranchName": "main",
          "S3": {
            "Bucket": "cdk-hnb659fds-assets-247574246160-ap-northeast-3",
            "Key": "018f30c2ae665c26fe4352d9e3116c29784b646ceb7b31decb3dd85b42000c58.zip",
          },
        },
        "RepositoryDescription": "Monorepo",
        "RepositoryName": "Monorepo",
      },
      "Type": "AWS::CodeCommit::Repository",
    },
    "CodeCommitMonorepoProjectStackCodePipeline3C086EC1mainEventRuleAB0B1DD8": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "event": [
              "referenceCreated",
              "referenceUpdated",
            ],
            "referenceName": [
              "main",
            ],
            "referenceType": [
              "branch",
            ],
          },
          "detail-type": [
            "CodeCommit Repository State Change",
          ],
          "resources": [
            {
              "Fn::GetAtt": [
                "CodeCommit506A881C",
                "Arn",
              ],
            },
          ],
          "source": [
            "aws.codecommit",
          ],
        },
        "Name": "triggerMonorepoPipeline-lambda-rule",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "LambdaD247545B",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CodeCommitMonorepoProjectStackCodePipeline3C086EC1mainEventRuleAllowEventRuleMonorepoProjectStackLambda8EEF25E61ABB6199": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "LambdaD247545B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "CodeCommitMonorepoProjectStackCodePipeline3C086EC1mainEventRuleAB0B1DD8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CodePipelineActionIAMRoleA42EA005": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::247574246160:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "codecommit:*",
                    "codebuild:*",
                    "codepipeline:*",
                    "logs:*",
                    "s3:*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "codepipelineActionPolicy",
          },
        ],
        "RoleName": "pipelineActionIAMRole-monorepo",
      },
      "Type": "AWS::IAM::Role",
    },
    "CodePipelineActionIAMRoleDefaultPolicy02298C9C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3CodePipelineArtifact91067FA4",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3CodePipelineArtifact91067FA4",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codecommit:GetBranch",
                "codecommit:GetCommit",
                "codecommit:UploadArchive",
                "codecommit:GetUploadArchiveStatus",
                "codecommit:CancelUploadArchive",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodeCommit506A881C",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodeBuild2FDE9E35",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodePipelineActionIAMRoleDefaultPolicy02298C9C",
        "Roles": [
          {
            "Ref": "CodePipelineActionIAMRoleA42EA005",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CodePipelineB74E5936": {
      "DependsOn": [
        "CodePipelineIAMRoleDefaultPolicyDEAAEF9D",
        "CodePipelineIAMRoleABCE4B1D",
      ],
      "Properties": {
        "ArtifactStore": {
          "Location": {
            "Ref": "S3CodePipelineArtifact91067FA4",
          },
          "Type": "S3",
        },
        "ExecutionMode": "QUEUED",
        "Name": "monorepo-pipeline",
        "PipelineType": "V2",
        "RoleArn": {
          "Fn::GetAtt": [
            "CodePipelineIAMRoleABCE4B1D",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeCommit",
                  "Version": "1",
                },
                "Configuration": {
                  "BranchName": "main",
                  "PollForSourceChanges": false,
                  "RepositoryName": {
                    "Fn::GetAtt": [
                      "CodeCommit506A881C",
                      "Name",
                    ],
                  },
                },
                "Name": "SourceCodeCommit",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_source_SourceCodeCommit",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "CodePipelineActionIAMRoleA42EA005",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "EnvironmentVariables": "[{"name":"sourceDirectoryName","type":"PLAINTEXT","value":"#{variables.sourceDirectoryName}"},{"name":"deployS3BucketName","type":"PLAINTEXT","value":"#{variables.deployS3BucketName}"}]",
                  "ProjectName": {
                    "Ref": "CodeBuild2FDE9E35",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Artifact_source_SourceCodeCommit",
                  },
                ],
                "Name": "BuildS3",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "CodePipelineActionIAMRoleA42EA005",
                    "Arn",
                  ],
                },
                "RunOrder": 2,
              },
            ],
            "Name": "Build",
          },
        ],
        "Variables": [
          {
            "DefaultValue": "default",
            "Description": "Receive from Lambda",
            "Name": "sourceDirectoryName",
          },
          {
            "DefaultValue": "default",
            "Description": "Receive from Lambda",
            "Name": "deployS3BucketName",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "CodePipelineIAMRoleABCE4B1D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSCodePipeline_FullAccess",
              ],
            ],
          },
        ],
        "RoleName": "pipelineIAMRole-monorepo",
      },
      "Type": "AWS::IAM::Role",
    },
    "CodePipelineIAMRoleDefaultPolicyDEAAEF9D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3CodePipelineArtifact91067FA4",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3CodePipelineArtifact91067FA4",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "CodePipelineActionIAMRoleA42EA005",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodePipelineIAMRoleDefaultPolicyDEAAEF9D",
        "Roles": [
          {
            "Ref": "CodePipelineIAMRoleABCE4B1D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaCloudWatchLogs2EB2B492": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupClass": "STANDARD",
        "LogGroupName": "triggerMonorepoPipeline-lambda-logs",
        "RetentionInDays": 1,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaD247545B": {
      "DependsOn": [
        "LambdaIAMRoleCA1E3376",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-247574246160-ap-northeast-3",
          "S3Key": "d3a06334ae220c0dfc44126afdaa22d24af6143f42b79ef384fef165c994c947.zip",
        },
        "Environment": {
          "Variables": {
            "CODEPIPELINE": "monorepo-pipeline",
            "LOG_LEVEL": "DEBUG",
            "S3_SYNC_DIR_LIST": "monorepo-target-example-s3-a,monorepo-target-example-s3-b",
          },
        },
        "FunctionName": "triggerMonorepoPipeline-lambda",
        "Handler": "index.lambda_handler",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "LambdaCloudWatchLogs2EB2B492",
          },
        },
        "Role": {
          "Fn::GetAtt": [
            "LambdaIAMRoleCA1E3376",
            "Arn",
          ],
        },
        "Runtime": "python3.12",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaIAMRoleCA1E3376": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "codecommit:*",
                    "codepipeline:*",
                    "logs:*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "codebuildPolicy",
          },
        ],
        "RoleName": "lambdaIAMRole-triggerMonorepoPipeline-lambda",
      },
      "Type": "AWS::IAM::Role",
    },
    "S3CodePipelineArtifact91067FA4": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "monorepo-pipeline-artifact",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "S3DeployTargetABCE05E21": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "monorepo-target-example-s3-a",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "S3DeployTargetB970BC0D5": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "monorepo-target-example-s3-b",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
